File: api/schemas/import_data.py
"""Esquemas de la importación de ensayos desde archivo."""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


class ImportedVariableSummary(BaseModel):
    name: str = Field(..., description="Nombre tal como aparece en el archivo")
    samples: int
    min: Optional[float]
    max: Optional[float]
    first: Optional[float]
    last: Optional[float]
    preview: list[float] = Field(..., description="Primeros valores, para el modal")
    constant: bool = Field(
        ..., description="True si la señal no se mueve: no sirve como actuador ni sensor"
    )


class ParseResponse(BaseModel):
    """Lo que devuelve /parse: qué trae el archivo, para armar el modal de mapeo."""

    ok: bool
    token: str = Field(
        ...,
        description=(
            "Identificador del archivo parseado. Se envía luego en /load para "
            "no volver a subir el archivo."
        ),
    )
    source_name: str
    format: str = Field(..., description="codesys-trace | csv | xlsx")
    samples: int
    duration_s: float
    sample_period_s: float
    variables: list[ImportedVariableSummary]
    suggested_mapping: dict[str, Optional[str]] = Field(
        ...,
        description=(
            "Sugerencia por heurística sobre los NOMBRES del archivo (busca "
            "'out'/'pid' para actuador, 'sp'/'setpoint' para consigna...). "
            "Solo pre-selecciona el modal; el usuario decide."
        ),
    )

    model_config = {
        "json_schema_extra": {
            "example": {
                "ok": True,
                "token": "a3f8c2",
                "source_name": "data_sinto_05_08.trace.csv",
                "format": "codesys-trace",
                "samples": 1517,
                "duration_s": 30.32,
                "sample_period_s": 0.02,
                "variables": [
                    {
                        "name": "Main_Control.PID_OUT",
                        "samples": 1517,
                        "min": 25.0,
                        "max": 50.0,
                        "first": 25.0,
                        "last": 50.0,
                        "preview": [25.0, 25.0, 25.0],
                        "constant": False,
                    }
                ],
                "suggested_mapping": {
                    "actuator": "Main_Control.PID_OUT",
                    "sensor": "Main_Control.Velocidad_Scaled",
                    "setpoint": "DB_HMI.HMI_SP_Local_Automatico",
                },
            }
        }
    }


class ImportMapping(BaseModel):
    """Qué variable del ARCHIVO cumple cada rol."""

    actuator: str = Field(..., description="Obligatoria: la entrada del proceso")
    sensor: str = Field(..., description="Obligatoria: la respuesta del proceso")
    setpoint: Optional[str] = Field(None, description="Opcional")


class LoadRequest(BaseModel):
    token: str = Field(..., description="El devuelto por /parse")
    mapping: ImportMapping
    scale: str = Field(
        "pct",
        description=(
            "Escala de las señales del archivo (ma | pct | v). Un trace de "
            "CODESYS suele venir en unidades de ingeniería 0-100: pct."
        ),
    )

    model_config = {
        "json_schema_extra": {
            "example": {
                "token": "a3f8c2",
                "mapping": {
                    "actuator": "Main_Control.PID_OUT",
                    "sensor": "Main_Control.Velocidad_Scaled",
                    "setpoint": "DB_HMI.HMI_SP_Local_Automatico",
                },
                "scale": "pct",
            }
        }
    }


class ImportStatus(BaseModel):
    source: str = Field(..., description="realtime | imported")
    active: bool = Field(..., description="True si el buffer contiene datos de archivo")
    source_name: Optional[str] = None
    format: Optional[str] = None
    samples: Optional[int] = None
    sample_period_s: Optional[float] = None
    mapping: Optional[dict] = None
    loaded_at: Optional[float] = None

    model_config = {
        "json_schema_extra": {
            "example": {
                "source": "imported",
                "active": True,
                "source_name": "data_sinto_05_08.trace.csv",
                "format": "codesys-trace",
                "samples": 1517,
                "sample_period_s": 0.02,
                "mapping": {
                    "actuator": "Main_Control.PID_OUT",
                    "sensor": "Main_Control.Velocidad_Scaled",
                    "setpoint": "DB_HMI.HMI_SP_Local_Automatico",
                },
                "loaded_at": 1754413200.5,
            }
        }
    }


class LoadResponse(ImportStatus):
    ok: bool = True
    buffer_size: int = 0

File: web/static/js/import_data.js
/* =========================================================
   import_data.js
   Importación de ensayos desde archivo (botón "Importar").

   Flujo:
     1. El usuario elige un archivo (.trace.csv de CODESYS, CSV o xlsx).
     2. POST /api/import/parse -> variables encontradas.
     3. Modal: asignar qué variable del ARCHIVO es actuador/sensor/setpoint.
     4. POST /api/import/load -> el buffer del backend pasa a ser el archivo
        (modo importado: las muestras del PLC se ignoran).
     5. Identificar / gráficos / PID funcionan igual que en vivo.

   El banner de modo importado y el botón "Volver a tiempo real" viven aquí.
   ========================================================= */


let _importParse = null;  // respuesta de /parse, viva mientras el modal está abierto


/** Entrada: el <input type="file"> oculto de la topbar dispara esto. */
async function handleImportFile(input) {
  const file = input.files && input.files[0];
  input.value = "";  // permite volver a elegir el mismo archivo
  if (!file) return;

  setStatus(`Analizando ${file.name}...`, "running");

  const form = new FormData();
  form.append("file", file, file.name);

  try {
    const response = await fetch(`${State.API_BASE}/api/import/parse`, {
      method: "POST",
      body: form
    });

    const data = await response.json().catch(() => ({}));
    if (!response.ok) throw new Error(data.detail || `HTTP ${response.status}`);

    _importParse = data;
    renderImportModal(data);
    setStatus(
      `${file.name}: ${data.samples} muestras · ${data.variables.length} variables · ` +
        `${(data.sample_period_s * 1000).toFixed(0)} ms`,
      "ok"
    );
  } catch (err) {
    console.error("Error al analizar el archivo:", err);
    setStatus(`No se pudo leer el archivo: ${err.message}`, "error");
  }
}


/* ==================== MODAL DE MAPEO ==================== */

function renderImportModal(data) {
  closeImportModal();

  const overlay = document.createElement("div");
  overlay.id = "importModal";
  overlay.className = "import-overlay";

  const options = (rol, opcional) => {
    const sugerido = data.suggested_mapping?.[rol] || "";
    const opts = opcional ? ['<option value="">— sin asignar —</option>'] : [];
    data.variables.forEach((v) => {
      const sel = v.name === sugerido ? " selected" : "";
      const aviso = v.constant ? " (constante)" : "";
      opts.push(
        `<option value="${escapeHtml(v.name)}"${sel}>` +
          `${escapeHtml(v.name)}${aviso}</option>`
      );
    });
    return opts.join("");
  };

  const filas = data.variables
    .map(
      (v) => `
      <tr>
        <td class="imp-name">${escapeHtml(v.name)}</td>
        <td>${v.samples}</td>
        <td>${formatNumber(v.min, 2)} … ${formatNumber(v.max, 2)}</td>
        <td>${v.constant ? '<span class="imp-warn">constante</span>' : "ok"}</td>
      </tr>`
    )
    .join("");

  overlay.innerHTML = `
    <div class="import-modal">
      <h3>Importar ensayo — ${escapeHtml(data.source_name)}</h3>
      <p class="imp-meta">
        ${data.format} · ${data.samples} muestras ·
        ${(data.duration_s).toFixed(1)} s · periodo ${(data.sample_period_s * 1000).toFixed(0)} ms
      </p>

      <table class="imp-table">
        <thead><tr><th>Variable del archivo</th><th>Muestras</th><th>Rango</th><th></th></tr></thead>
        <tbody>${filas}</tbody>
      </table>

      <div class="imp-grid">
        <div>
          <label>Entrada (u) — Actuador*</label>
          <select id="impAct">${options("actuator", false)}</select>
        </div>
        <div>
          <label>Salida (y) — Sensor*</label>
          <select id="impSen">${options("sensor", false)}</select>
        </div>
        <div>
          <label>Set Point — opcional</label>
          <select id="impSP">${options("setpoint", true)}</select>
        </div>
        <div>
          <label>Escala de las señales</label>
          <select id="impScale">
            <option value="pct" selected>0-100 % (unidades de ingeniería)</option>
            <option value="ma">4-20 mA</option>
            <option value="v">0-10 V</option>
          </select>
        </div>
      </div>

      <div class="imp-actions">
        <button class="btn btn-secondary" onclick="closeImportModal()">Cancelar</button>
        <button class="btn btn-primary" onclick="confirmImport()">
          <i class="fa-solid fa-file-import"></i> Cargar datos
        </button>
      </div>
    </div>
  `;

  document.body.appendChild(overlay);
}


function closeImportModal() {
  document.getElementById("importModal")?.remove();
}


async function confirmImport() {
  if (!_importParse) return;

  const actuator = document.getElementById("impAct")?.value || "";
  const sensor   = document.getElementById("impSen")?.value || "";
  const setpoint = document.getElementById("impSP")?.value || null;
  const scale    = document.getElementById("impScale")?.value || "pct";

  if (!actuator || !sensor) {
    setStatus("Asigna al menos actuador y sensor.", "error");
    return;
  }
  if (actuator === sensor) {
    setStatus("Actuador y sensor no pueden ser la misma variable.", "error");
    return;
  }

  setStatus("Cargando datos importados...", "running");

  try {
    const response = await fetch(`${State.API_BASE}/api/import/load`, {
      method:  "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        token: _importParse.token,
        mapping: { actuator, sensor, setpoint },
        scale
      })
    });

    const data = await response.json().catch(() => ({}));
    if (!response.ok) throw new Error(data.detail || `HTTP ${response.status}`);

    closeImportModal();
    _importParse = null;

    enterImportedMode(data);

    // El buffer del backend ya es el archivo: pedir la serie y volcarla al
    // sampleStore para que los gráficos del paso 3 la muestren tal cual.
    await loadImportedSeriesIntoStore();

    setStatus(
      `Datos importados: ${data.source_name} (${data.samples} muestras). ` +
        `Ve al paso 3 para ver las señales, o Identificar directamente.`,
      "ok"
    );
  } catch (err) {
    console.error("Error al cargar importado:", err);
    setStatus(`No se pudo cargar: ${err.message}`, "error");
  }
}


/**
 * Trae la serie completa del backend y la vuelca al sampleStore local.
 * Así plotCapture y los textareas muestran el archivo sin tocar su código.
 */
async function loadImportedSeriesIntoStore() {
  try {
    const response = await fetch(`${State.API_BASE}/api/test/series?percent=false`);
    if (!response.ok) return;
    const serie = await response.json();
    fillStoreFromSeries(serie);
  } catch (_) {
    // Fallback por WebSocket si el endpoint REST no está.
    sendWsMessage({ type: "get_series" });
  }
}


function fillStoreFromSeries(serie) {
  if (!serie || !Array.isArray(serie.time)) return;

  resetSampleStore();
  const s = State.sampleStore;

  serie.time.forEach((t, i) => {
    s.time.push(t);
    s.actuator_ma.push(serie.actuator?.[i] ?? null);
    s.sensor_ma.push(serie.sensor?.[i] ?? null);
    s.setpoint_ma.push(serie.setpoint?.[i] ?? null);
    // En importado la escala cruda YA es la de trabajo; los _pct los calcula
    // el backend en cada sample, aquí se replican para el toggle de la vista.
    s.actuator_pct.push(serie.actuator?.[i] ?? null);
    s.sensor_pct.push(serie.sensor?.[i] ?? null);
    s.setpoint_pct.push(serie.setpoint?.[i] ?? null);
    s.actuator_cmd_ma.push(null);
    s.actuator_cmd_pct.push(null);
  });

  // Eje X del gráfico = duración del archivo.
  State.ensayo.durationS = serie.time.length ? serie.time[serie.time.length - 1] : null;
  State.ensayo.plan = null;  // no hay línea objetivo: esto no es un ensayo del runner

  if (typeof plotCapture === "function") plotCapture();
  if (typeof fillManualTextareas === "function") fillManualTextareas();
}


/* ==================== BANNER DE MODO IMPORTADO ==================== */

function enterImportedMode(info) {
  removeImportBanner();

  const banner = document.createElement("div");
  banner.id = "importBanner";
  banner.className = "import-banner";
  banner.innerHTML = `
    <i class="fa-solid fa-file-import"></i>
    <span>
      <strong>Datos importados:</strong> ${escapeHtml(info.source_name || "archivo")}
      · ${info.samples} muestras · ${((info.sample_period_s || 0) * 1000).toFixed(0)} ms
      — las señales del PLC en vivo están pausadas
    </span>
    <button class="btn btn-secondary btn-sm" onclick="exitImportedMode()">
      Volver a tiempo real
    </button>
  `;

  const topbar = document.querySelector(".topbar");
  (topbar?.parentNode || document.body).insertBefore(banner, topbar?.nextSibling || null);

  // Durante el modo importado no tiene sentido arrancar un ensayo del runner.
  const btnStart = document.getElementById("btnStart");
  if (btnStart) btnStart.style.display = "none";
  const btnIdent = document.getElementById("btnIdent");
  if (btnIdent) btnIdent.style.display = "";
}


function removeImportBanner() {
  document.getElementById("importBanner")?.remove();
}


async function exitImportedMode() {
  try {
    await fetch(`${State.API_BASE}/api/import/clear`, { method: "POST" });
  } catch (_) {}

  removeImportBanner();
  resetSampleStore();
  if (typeof plotCapture === "function") plotCapture();

  const btnStart = document.getElementById("btnStart");
  if (btnStart) btnStart.style.display = "";

  setStatus("Modo importado cerrado. Se aceptan de nuevo señales del PLC.", "ok");
}


/** Al cargar la vista: si el backend sigue en modo importado, reflejarlo. */
async function refreshImportState() {
  try {
    const response = await fetch(`${State.API_BASE}/api/import/status`);
    if (!response.ok) return;
    const data = await response.json();

    if (data.active) {
      enterImportedMode(data);
      await loadImportedSeriesIntoStore();
    }
  } catch (_) {}
}
